std/fmt/parse_float.zig

```diff
@@ -0,0 +1,425 @@
+                        n.exponent += 1;
+                    }
+
+                    i += 1;
+                } else if (c == '.') {
+                    i += 1;
+                    state = State.MantissaFractional;
+                } else {
+                    state = State.MantissaFractional;
+                }
+            },
+
+            State.MantissaFractional => {
+                if (isDigit(c)) {
+                    if (digit_index < max_digits) {
+                        n.mantissa *%= 10;
+                        n.mantissa += c - '0';
+                        n.exponent -%= 1;
+                        digit_index += 1;
+                    }
+
+                    i += 1;
+                } else if (c == 'e' or c == 'E') {
+                    i += 1;
+                    state = State.ExponentSign;
+                } else {
+                    state = State.ExponentSign;
+                }
+            },
+
+            State.ExponentSign => {
+                if (c == '+') {
+                    i += 1;
+                } else if (c == '-') {
+                    negative_exp = true;
+                    i += 1;
+                }
+
+                state = State.LeadingExponentZeros;
+            },
+
+            State.LeadingExponentZeros => {
+                if (c == '0') {
+                    i += 1;
+                } else {
+                    state = State.Exponent;
+                }
+            },
+
+            State.Exponent => {
+                if (isDigit(c)) {
+                    if (exponent < std.math.maxInt(i32)) {
+                        exponent *= 10;
+                        exponent += @intCast(i32, c - '0');
+                    }
+
+                    i += 1;
+                } else {
+                    state = State.Stop;
+                }
+            },
+
+            State.Stop => break :loop,
+        }
+    }
+
+    if (negative_exp) exponent = -exponent;
+    n.exponent += exponent;
+
+    if (n.mantissa == 0) {
+        return if (n.negative) ParseResult.MinusZero else ParseResult.PlusZero;
+    } else if (n.exponent > 309) {
+        return if (n.negative) ParseResult.MinusInf else ParseResult.PlusInf;
+    } else if (n.exponent < -328) {
+        return if (n.negative) ParseResult.MinusZero else ParseResult.PlusZero;
+    }
+
+    return ParseResult.Ok;
+}
+
+inline fn isLower(c: u8) bool {
+    return c -% 'a' < 26;
+}
+
+inline fn toUpper(c: u8) u8 {
+    return if (isLower(c)) (c & 0x5f) else c;
+}
+
+fn caseInEql(a: []const u8, b: []const u8) bool {
+    if (a.len != b.len) return false;
+
+    for (a) |_, i| {
+        if (toUpper(a[i]) != toUpper(b[i])) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+pub fn parseFloat(comptime T: type, s: []const u8) T {
+    var r = FloatRepr{
+        .negative = false,
+        .exponent = 0,
+        .mantissa = 0,
+    };
+
+    if (caseInEql(s, "nan")) {
```

Comment:
Would be worthwhile to add an extra condition here to avoid 4 compares in the non-special case.

Reply:
From my experience doing anything smart around string comparisons tend to not be worth it as the compiler is pretty good in this area. https://github.com/Hejsil/fun-with-zig/blob/e8f857a401151dbb716acce1775e83c4d1656132/bench/match.zig#L9-L18

Reply:
Good point. Let's measure it!

https://zig.godbolt.org/z/9zC4yU

The slow-case is actually far worse than the 4 I said since there are a lot of branches in the slice equality check. Without the check we have to go through most of the branches for any value 3 or 4 digits in length. Definitely think this is worth it given the float-parsing code is often the bottleneck in high-performance parsing.